### PR TITLE
Bridge gas optimisations

### PR DESCRIPTION
**Pull request by eth-r**
The bridge is mostly fairly close to the limit of achievable performance, but a few optimisations could be identified:

- some hashing operations can be made more efficient
- storage slots in `BridgeState` can be aligned and packed so that multiple accesses are minimised
- some allocations can be avoided by adding at-functions in `bitcoin-spv`

All in all, single percent improvements can be made in some functions (particularly proof submissions) and rarely called operations involving storage (initialisation, updating parameters) become significantly cheaper

Refs #260

### CODE REVIEW

**Comment by pdyraga** [comment]
Let's also inspect the usage of `bytes memory` across the project. I was able to get some small savings for `Bank.approveBalanceAndCall` in 237b3c1013dce709025772af2dc3888e92b8480a (PR #302) but I haven't gone through all `bytes memory` usages in the project to not run into conflicts with this PR.

**Comment by pdyraga** [comment]
Let's also inspect the usage of `bytes memory` across the project. I was able to get some small savings for `Bank.approveBalanceAndCall` in 237b3c1013dce709025772af2dc3888e92b8480a (PR #302) but I haven't gone through all `bytes memory` usages in the project to not run into conflicts with this PR.

File: solidity/contracts/bridge/BridgeState.sol [hunk 1]
```diff
@@ -61,20 +61,8 @@ library BridgeState {
...
-        mapping(uint256 => Deposit.DepositRequest) deposits;
-        // Indicates if the vault with the given address is trusted or not.
-        // Depositors can route their revealed deposits only to trusted vaults
-        // and have trusted vaults notified about new deposits as soon as these
-        // deposits get swept. Vaults not trusted by the Bridge can still be
-        // used by Bank balance owners on their own responsibility - anyone can
-        // approve their Bank balance to any address.
-        mapping(address => bool) isVaultTrusted;
+        // slither-disable-next-line unused-state
+        bytes8 __depositAlignmentGap;
```
Comment (tomaszslabon) [inline solidity/contracts/bridge/BridgeState.sol:65]: Could we add a comment why we added this variable?
(I'm guessing it's to make the layout of variables more efficient).

File: solidity/contracts/bridge/BridgeState.sol [hunk 2]
```diff
@@ -162,68 +143,31 @@ library BridgeState {
         // This is a per-redemption output max fee for the redemption
         // transaction.
         uint64 redemptionTxMaxFee;
+        // slither-disable-next-line unused-state
+        bytes8 __redemptionAlignmentGap;
```
Comment (tomaszslabon) [inline solidity/contracts/bridge/BridgeState.sol:147]: Could we add a comment why we added this variable?

Comment (tomaszslabon) [inline solidity/contracts/bridge/BridgeState.sol:147]: Could we add a comment explaining why we added this variable, something like:
// for a more efficient variable layout

**Comment by pdyraga** [comment]
👀

**Review by tomaszslabon — APPROVED** [review]

**Review by pdyraga — APPROVED** [review]

File: solidity/contracts/bridge/BitcoinTx.sol [hunk 3]
```diff
@@ -278,49 +278,82 @@ library BitcoinTx {
...
-        // - 0x00: OP_0
-        // - 0x14: Byte length of the public key hash
-        // which matches the P2WPKH structure as per:
-        // https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#P2WPKH
-        bytes32 P2WPKHScriptKeccak = keccak256(
-            abi.encodePacked(hex"160014", pubKeyHash)
-        );
-        // Make sure the actual output script matches either the P2PKH
-        // or P2WPKH format.
+        uint256 scriptLen = output.length - 8;
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/BitcoinTx.sol:281]: Addressed in #345 in https://github.com/keep-network/tbtc-v2/pull/345/commits/9aeb58cfb338a92e0ef4265538890799479d7158.

File: solidity/contracts/bridge/BridgeState.sol [hunk 4]
```diff
@@ -61,20 +61,10 @@ library BridgeState {
...
-        // Depositors can route their revealed deposits only to trusted vaults
-        // and have trusted vaults notified about new deposits as soon as these
-        // deposits get swept. Vaults not trusted by the Bridge can still be
-        // used by Bank balance owners on their own responsibility - anyone can
-        // approve their Bank balance to any address.
-        mapping(address => bool) isVaultTrusted;
+        // Move movingFundsTxMaxTotalFee to the next storage slot for a more
+        // efficient variable layout
+        // slither-disable-next-line unused-state
+        bytes8 __depositAlignmentGap;
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/BridgeState.sol:67]: Addressed in #345 in https://github.com/keep-network/tbtc-v2/pull/345/commits/9d15d6e57956481b7ee007a339f595482b22f8d3.

File: solidity/contracts/bridge/BridgeState.sol [hunk 5]
```diff
@@ -162,68 +145,33 @@ library BridgeState {
...
+        // variable layout
+        // slither-disable-next-line unused-state
+        bytes8 __redemptionAlignmentGap;
         // Time after which the redemption request can be reported as
         // timed out. It is counted from the moment when the redemption
         // request was created via `requestRedemption` call. Reported
         // timed out requests are cancelled and locked TBTC is returned
         // to the redeemer in full amount.
-        uint256 redemptionTimeout;
+        uint64 redemptionTimeout;
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/BridgeState.sol:157]: Addressed in #345 in https://github.com/keep-network/tbtc-v2/pull/345/commits/9d15d6e57956481b7ee007a339f595482b22f8d3.

File: solidity/contracts/bridge/BridgeState.sol [hunk 6]
```diff
@@ -129,14 +119,7 @@ library BridgeState {
         // The percentage of the notifier reward from the staking contract
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/BridgeState.sol:119]: Addressed in #345 in https://github.com/keep-network/tbtc-v2/pull/345/commits/9d15d6e57956481b7ee007a339f595482b22f8d3.

File: solidity/contracts/bridge/Bridge.sol [hunk 7]
```diff
@@ -1228,9 +1228,9 @@ contract Bridge is
         uint64 redemptionDustThreshold,
         uint64 redemptionTreasuryFeeDivisor,
         uint64 redemptionTxMaxFee,
-        uint256 redemptionTimeout,
+        uint64 redemptionTimeout,
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/Bridge.sol:1231]: Addressed in #345 in https://github.com/keep-network/tbtc-v2/pull/345/commits/a3716d08d9049fa475f1570a6c74ab7d566879a4.

File: solidity/contracts/bridge/Redemption.sol [hunk 8]
```diff
@@ -1017,4 +1013,43 @@ library Redemption {
         // Return the requested amount of tokens to the redeemer
         self.bank.transferBalance(request.redeemer, request.requestedAmount);
     }
+
+    /// @notice Calculate redemption key without allocations.
+    /// @param walletPubKeyHash the pubkey hash of the wallet.
+    /// @param script the output script of the redemption.
+    /// @return The key = keccak256(keccak256(script), walletPubKeyHash).
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/Redemption.sol:1020]: Addressed in #345 in https://github.com/keep-network/tbtc-v2/pull/345/commits/73dee0bafb727fcdb4ee499b0fb1c4f73b8db8be.

File: solidity/contracts/bridge/Redemption.sol [hunk 9]
```diff
@@ -783,24 +782,33 @@ library Redemption {
...
             // 9th byte until the end.
-            bytes memory outputScript = output.slice(8, output.length - 8);
+
+            uint256 outputScriptStart = processInfo.outputStartingIndex + 8;
+
+            bytes32 outputScriptHash;
+            /* solhint-disable-next-line no-inline-assembly */
+            assembly {
+                outputScriptHash := keccak256(
+                    add(redemptionTxOutputVector, add(outputScriptStart, 32)),
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/Redemption.sol:803]: Addressed in #345 in https://github.com/keep-network/tbtc-v2/pull/345/commits/01b490cfd329520026b691e8ddc8958353adc600.

File: solidity/contracts/bridge/Redemption.sol [hunk 10]
```diff
@@ -783,24 +782,33 @@ library Redemption {
             //       https://github.com/keep-network/tbtc-v2/issues/257
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/Redemption.sol:782]: Addressed in #345 in https://github.com/keep-network/tbtc-v2/pull/345/commits/7082d42267d12b21a2d912490b086ade4b2873b3.

File: solidity/contracts/bridge/Redemption.sol [hunk 11]
```diff
@@ -783,24 +782,33 @@ library Redemption {
...
-                outputLength
-            );
 
             // Extract the value from given output.
-            uint64 outputValue = output.extractValue();
+            uint64 outputValue = redemptionTxOutputVector.extractValueAt(
+                processInfo.outputStartingIndex
+            );
+
+            uint256 scriptLength = outputLength - 8;
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/Redemption.sol:791]: Addressed in #345 in https://github.com/keep-network/tbtc-v2/pull/345/commits/5531f8f50bb5dea50a76e1b36ef62f4b12989cd9.

File: solidity/contracts/bridge/Redemption.sol [hunk 12]
```diff
@@ -864,17 +868,9 @@ library Redemption {
...
     function processNonChangeRedemptionTxOutput(
         BridgeState.Storage storage self,
-        bytes20 walletPubKeyHash,
-        bytes memory outputScript,
+        uint256 redemptionKey,
         uint64 outputValue
     ) internal returns (uint64 burnableValue, uint64 treasuryFee) {
-        // This function should be called only if the given output is
-        // supposed to represent a redemption. Build the redemption key
-        // to perform that check.
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/Redemption.sol:873]: Addressed in #345 in https://github.com/keep-network/tbtc-v2/pull/345/commits/f6a1740c795a0f46c8aba0b167538d3e37f0e028.

**Review by pdyraga — APPROVED** [review]

**Pull request merged**